### PR TITLE
feat: support `ChaCha20RNG`

### PR DESCRIPTION
**Pull request by chokobole**
# Description

Scroll's `zkevm-circuit` proof generation is delegated to https://github.com/scroll-tech/snark-verifier. Since this depends on `ChaCha20RNG`, this PR is made.

### CODE REVIEW

**Review by dongchangYoo — APPROVED** [review]
LGTM

File: tachyon/crypto/random/rng_type.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+#include "tachyon/base/flag/flag_value_traits.h"
+
+namespace tachyon {
+namespace crypto {
+
+// THE ORDER OF ELEMENTS ARE IMPORTANT!! DO NOT CHANGE!
+// This order matches with the constants of c APIs.
+// Pleas refer to tachyon/c/crypto/random.h for details.
+enum class RngType : uint8_t {
```
Comment (TomTaehoonKim) [inline tachyon/crypto/random/rng_type.h:18]:
```suggestion
enum class RNGType : uint8_t {
```
Optional comment. You can resolve it yourself if you think it's not needed.

File: vendors/halo2/src/cha_cha20_rng.cc [hunk 2]
```diff
@@ -0,0 +1,48 @@
+#include "vendors/halo2/include/cha_cha20_rng.h"
+
+#include <string.h>
```
Comment (ashjeong) [inline vendors/halo2/src/cha_cha20_rng.cc:3]: Where is this used?

File: vendors/circom/README.md [hunk 3]
```diff
@@ -2,27 +2,30 @@
...
+| Field Support             | All       | Bn128           | Bn128                             |
+| [FFIASM] Generated Field  | Yes(Fast) | No(Slow)        | Yes(Fast)                         |
+| Gpu Support               | Yes       | No              | Yes([Orbiter-Finance/Rapidsnark]) |
 
 1. Speed: Tachyon runs approximately 10 times faster than [Rapidsnark]! The [benchmark] was conducted using [sha256_512.circom] with a degree of 2¹⁶.
 2. Seamless Workflow: Tachyon doesn't require you to run any additional programs just to generate a circom proof. No need for [snarkjs] for witness file generation.
 3. Flexible Witness Handling: With Tachyon, you can modify the witness directly in your program. No need to create a separate JSON file with [snarkjs].
 4. Integrated Build System: Tachyon seamlessly integrates circom compilation into the build system, specifically bazel. When you make changes to a circom file, it's compiled accordingly and built in parallel, ensuring a safe and efficient process.
 5. Field Support: Tachyon isn't limited to bn128; it easily supports all fields!
+6. Gpu Support: Both Tachyon and [Orbiter-Finance/Rapidsnark] support MSM / NTT support using [icicle].
```
Comment (ashjeong) [inline vendors/circom/README.md:19]:
```suggestion
6. Gpu Support: Both Tachyon and [Orbiter-Finance/Rapidsnark] support MSM / NTT using [icicle].

**Review by ashjeong — COMMENTED** [review]
2ccfcbb8407d96e5a437c8beb8f4cbe8a453c315
Please fix the type of the commit title to "description"

File: vendors/halo2/src/cha_cha20_rng.cc (hunk 2 above)
Comment (chokobole) [reply vendors/halo2/src/cha_cha20_rng.cc:3]: It's for `memcpy`

**Review by ashjeong — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: tachyon/crypto/random/xor_shift/xor_shift_rng.h [hunk 4]
```diff
@@ -29,38 +28,28 @@ class XORShiftRNG final : public RNG {
...
-  }
-
-  static XORShiftRNG FromRandomSeed() {
+  // RNG methods
+  void SetRandomSeed() override {
     uint8_t seed[kSeedSize];
     uint64_t lo = base::Uniform(base::Range<uint64_t>::All());
     uint64_t hi = base::Uniform(base::Range<uint64_t>::All());
     memcpy(&seed[0], &lo, sizeof(uint64_t));
     memcpy(&seed[8], &hi, sizeof(uint64_t));
```
Comment (batzor) [inline tachyon/crypto/random/xor_shift/xor_shift_rng.h:38]:
```suggestion
    *reinterpret_cast<uint64_t*>(&seed[0]) =
        base::Uniform(base::Range<uint64_t>::All());
    *reinterpret_cast<uint64_t*>(&seed[8]) =
        base::Uniform(base::Range<uint64_t>::All());
```

It is minor(optional) but we can avoid `memcpy` with something like this.

**Review by GideokKim — APPROVED** [review]
LGTM

File: tachyon/crypto/random/xor_shift/xor_shift_rng.h (hunk 4 above)
Comment (batzor) [reply tachyon/crypto/random/xor_shift/xor_shift_rng.h:38]: Feel free to resolve once you read.

**Review by batzor — APPROVED** [review]
LGTM

File: tachyon/crypto/random/xor_shift/xor_shift_rng.h (hunk 4 above)
Comment (batzor) [reply tachyon/crypto/random/xor_shift/xor_shift_rng.h:38]: For curiosity I checked and the asm difference disappears from `-O1` https://godbolt.org/z/K481zYasf

**Pull request merged**